SQLQuery_Week09_Homework.sql

```diff
@@ -0,0 +1,414 @@
+FROM Books b
+    INNER JOIN Authors a ON a.Id = b.AuthorId
+GROUP BY a.Email
+HAVING COUNT(AuthorId) > 3;
+
+-- 5. Display the library name and the book title of the largest sale (the highest number of copies sold in one sale).
+
+GO
+CREATE VIEW BookIdAndNumberOfCopiesByLibrary
+AS
+    SELECT b.Title, s.LibraryId, s.NumberOfCopies
+    FROM Books_Sales bs
+     INNER JOIN Books b ON b.Id = bs.BookId
+     INNER JOIN Sales s ON s.Id = bs.SalesId;
+GO
+
+SELECT SUM(bv.NumberOfCopies), l.Name
+FROM BookIdAndNumberOfCopiesByLibrary bv
+    INNER JOIN Libraries l ON l.Id = bv.LibraryId
+GROUP BY l.Name
+HAVING SUM(bv.NumberOfCopies) = (SELECT MAX(NumberOfCopies) FROM BookIdAndNumberOfCopiesByLibrary)
+
+-- 6. Display all the librarians that are not on holiday and are hired more than 5 years ago.
+
+SELECT * 
+FROM Librarians l
+WHERE l.IsOnHoliday = 0 AND DATEDIFF(YYYY,l.HireDate,GETDATE()) > 5;
+
+--OR
+
+-- GO
+-- CREATE PROCEDURE GetExperienceByYearsForPresentLibrarians(
+--     @yearsExperience INT
+-- )
+-- AS 
+-- BEGIN
+--     SELECT *
+--     FROM Librarians l
+--     WHERE l.IsOnHoliday = 0 AND DATEDIFF(YYYY,l.HireDate,GETDATE()) > @yearsExperience;
+-- END
+-- GO
+
+-- EXEC GetExperienceByYearsForPresentLibrarians 5;
+
+-- 7. Create a stored procedure that displays, based on the book title, the number of libraries in which the book is found. Call the procedure with three different book titles.
+
+GO
+CREATE PROCEDURE GetNumberOfLibrariesByBookTitle(
+    @customerBookTitle VARCHAR(100)
+)
+AS 
+BEGIN
+    SELECT COUNT(l.Id)
+    FROM Books_Libraries bl
+        INNER JOIN Books b ON bl.BookId = b.Id
+        INNER JOIN Libraries l ON bl.LibraryId = l.Id
+    GROUP BY b.Title
+    HAVING b.Title = @customerBookTitle;
+END
+GO
+
+EXEC GetNumberOfLibrariesByBookTitle 'Book1';
+EXEC GetNumberOfLibrariesByBookTitle 'Book3';
+EXEC GetNumberOfLibrariesByBookTitle 'Book5';
+
+-- 8. Create a function that returns, based on the author's email, the number of copies sold for each of his / her books. Call the function with three different emails.
+
+GO
+CREATE VIEW BookIdAndNumberOfCopiesByAuthorId
+AS
+    SELECT b.Title, s.LibraryId, s.NumberOfCopies, b.AuthorId 
+    FROM Books_Sales bs
+     INNER JOIN Books b ON b.Id = bs.BookId
+     INNER JOIN Sales s ON s.Id = bs.SalesId;
+GO
+
+GO
+CREATE FUNCTION dbo.GetNumberOfCopiesOfBooksByAuthorsEmail(
+    @customerAuthorsEmail VARCHAR(100)
+)
+RETURNS INT
+AS
+BEGIN
+    DECLARE @numberOfCopies INT;
+    SET @numberOfCopies = (SELECT SUM(bc.NumberOfCopies)
+                           FROM BookIdAndNumberOfCopiesByAuthorId bc
+                                INNER JOIN Authors a ON a.Id = bc.AuthorId
+                            GROUP BY a.Email
+                            HAVING a.Email = @customerAuthorsEmail);
+    RETURN @numberOfCopies;
+END;
+GO
+
+SELECT dbo.GetNumberOfCopiesOfBooksByAuthorsEmail('mp@email.com') as 'Number of copies';
```

Comment:
Love the original way to solve this. I will provide it to your colleagues as an example so they can also see it and benefit from it. Hope you don't mind.